src/engines/hugin-inference-engine.ts

```diff
@@ -0,0 +1,226 @@
+  }
+
+  // A helper method for resetting the cached potentials and marginals.
+  private resetCache = () => {
+    this._potentials = {}
+    this._marginals = {}
+  }
+
+  // Implementation of various member functions
+  hasVariable = (name: string) => this._network[name] != null;
+  getVariables = () => Object.keys(this._network);
+
+  hasLevel = (name: string, level: string) => {
+    if (!this._network[name]) return false
+    const found = this._network[name].states.filter(x => x === level)
+    return found !== []
+  }
+
+  getLevels = (name: string) => {
+    if (!this._network[name]) return []
+    return [...this._network[name].states]
+  }
+
+  hasParent = (name: string, parent: string) => {
+    if (!this._network[name]) return false
+    const found = this._network[name].parents.filter(x => x === parent)
+    return found !== []
+  }
+
+  getParents = (name: string) => {
+    if (!this._network[name]) return []
+    return [...this._network[name].parents]
+  }
+
+  // NOTE: in order to prevent external mutation of the CPT, the
+  // result is cloned prior to returning it.
+  getDistribution = (name: string) => clone(this._network[name].cpt)
+
+  // NOTE: in order to prevent external mutation of the CPT, the
+  // input is cloned prior to assigining to the internally cached value.
+  setDistribution = (name: string, cpt: ICptWithParents | ICptWithoutParents) => {
+    const node: INode = this._network[name]
+    const expectedLevels = node.states
+    const expectedParents = node.parents
+
+    // Perform some sanity checks.
+    const err = (reason: string) => {
+      throw new Error(`Cannot set the distribution for ${name}.  ${reason}.`)
+    }
+    if (!node) err('The variable does not exist in the network')
+
+    const observedLevels = Array.isArray(cpt)
+      ? cpt.length > 0
+        ? Object.keys(cpt[0].then)
+        : []
+      : Object.keys(cpt).sort()
+
+    const observedParents = Array.isArray(cpt)
+      ? cpt.length > 0
+        ? Object.keys(cpt[0].when)
+        : []
+      : []
+
+    const hasCorrectLevels: boolean = observedLevels.every(x => expectedLevels.includes(x)) && expectedLevels.every(x => observedLevels.includes(x))
+    const hasCorrectParents: boolean = observedParents.every(x => expectedParents.includes(x)) && observedParents.every(x => expectedParents.includes(x))
+    if (!hasCorrectLevels) err('The provided distribution did not have the expected levels')
+    if (!hasCorrectParents) err('The provided distribution did not have the expected parents')
+
+    // Passed the sanity checks.   reset the cache and set the (cloned) value.
+    this.resetCache()
+    this._network[name].cpt = clone(cpt)
+  }
+
+  // Get the clique potentials.   If the internal cache is not empty, then
+  // return the cached values, otherwise, compute them with the current
+  // evidence
+  private getCliquesPotentials: () => ICliquePotentials = () => {
+    if (Object.keys(this._potentials).length === 0) {
```

Comment:
Or you can use `isEmpty` from `ramda` here
It also check object: `isEmpty({}) === true`

```suggestion
    if (isEmpty(this._potentials)) {
```